src/data/setoid/partition.lean

```diff
@@ -201,3 +212,131 @@ _ (subtype (@is_partition α)) _ (partial_order.to_preorder _) $ partition.rel_i
 end partition
 
 end setoid
+
+/-- A partition of a type index by another type. -/
+@[nolint has_inhabited_instance]
+structure indexed_partition {ι α : Type*} (s : ι → set α) :=
+(eq_of_mem : ∀ {x i j}, x ∈ s i → x ∈ s j → i = j)
+(some : ι → α)
+(some_mem : ∀ i, some i ∈ s i)
+(index : α → ι)
+(mem_index : ∀ x, x ∈ s (index x))
+
+/-- The non-constructive constructor for `indexed_partition`. -/
+noncomputable
+def indexed_partition.mk' {ι α : Type*} (s : ι → set α) (dis : ∀ i j, i ≠ j → disjoint (s i) (s j))
+(empty : ∀ i, (s i).nonempty) (ex : ∀ x, ∃ i, x ∈ s i) : indexed_partition s :=
+{ eq_of_mem := begin
+    classical,
+    intros x i j hxi hxj,
+    by_contra h,
+    exact dis _ _ h ⟨hxi, hxj⟩
+  end,
+  some := λ i, (empty i).some,
+  some_mem := λ i, (empty i).some_spec,
+  index := λ x, (ex x).some,
+  mem_index := λ x, (ex x).some_spec }
+
+namespace indexed_partition
+
+open set
+
+variables {ι α : Type*} {s : ι → set α} (hs : indexed_partition s)
+include hs
+
+lemma exists_mem (x : α) : ∃ i, x ∈ s i := ⟨hs.index x, hs.mem_index x⟩
+
+lemma Union : (⋃ i, s i) = univ :=
+by { ext x, simp [hs.exists_mem x] }
+
+lemma disjoint : ∀ {i j}, i ≠ j → disjoint (s i) (s j) :=
+λ i j h x ⟨hxi, hxj⟩,h (hs.eq_of_mem hxi hxj)
+
+lemma eq_index_of_mem {x i} (hxi : x ∈ s i) : i = hs.index x :=
+hs.eq_of_mem hxi (hs.mem_index x)
+
+/-- The equivalence relation associated to an indexed partition. Two
+elements are equivalent if they belong to the same set of the partition. -/
+protected def setoid (hs : indexed_partition s) : setoid α :=
+{ r := λ x y, ∃ i, x ∈ s i ∧ y ∈ s i,
+  iseqv := ⟨λ x, ⟨hs.index x, hs.mem_index x, hs.mem_index x⟩,
+            λ x y ⟨i, hxi, hyi⟩, ⟨i, hyi, hxi⟩,
+            λ x y z ⟨i, hxi, hyi⟩ ⟨j, hyj, hzj⟩, ⟨i, hxi, (hs.eq_of_mem hyj hyi) ▸ hzj⟩⟩ }
+
+lemma some_index (x : α) : hs.setoid.rel (hs.some (hs.index x)) x :=
```

Comment:
Can you also add `∀ x, index (some x) = x` or better the defeq `left_inverse index some`

Reply:
I don't think the `left_inverse` version is better because it prevents `simp` from using the lemma.